src/bin/cargo-fmt.rs

```diff
@@ -139,28 +154,113 @@ pub struct Target {
+            _ => false,
+        }
+    }
+
+    pub fn get_some<'a>(&'a self) -> Option<&'a [String]> {
+        use std::borrow::Borrow;
+        match self {
+            &WorkspaceHitlist::Some(ref hitlist) => Some(hitlist.borrow()),
+            _ => None,
+        }
+    }
+}
+
 // Returns a vector of all compile targets of a crate
-fn get_targets() -> Result<Vec<Target>, std::io::Error> {
+fn get_targets(workspace_hitlist: WorkspaceHitlist) -> Result<Vec<Target>, std::io::Error> {
     let mut targets: Vec<Target> = vec![];
-    let output = try!(Command::new("cargo").arg("read-manifest").output());
+    if workspace_hitlist.is_none() {
+        let output = try!(Command::new("cargo").arg("read-manifest").output());
+        if output.status.success() {
+            // None of the unwraps should fail if output of `cargo read-manifest` is correct
+            let data = &String::from_utf8(output.stdout).unwrap();
+            let json = Json::from_str(data).unwrap();
+            let jtargets = json.find("targets")
+                .unwrap()
+                .as_array()
+                .unwrap();
+            for jtarget in jtargets {
+                targets.push(target_from_json(jtarget));
+            }
+
+            return Ok(targets);
+        }
+        return Err(std::io::Error::new(std::io::ErrorKind::NotFound,
+                                       str::from_utf8(&output.stderr).unwrap()));
+    }
+    // This happens when cargo-fmt is not used inside a crate or
+    // is used inside a workspace.
+    // To ensure backward compatability, we only use `cargo metadata` for workspaces.
+    // TODO: How do we make sure we use either metadata or read-manifest
+    let output = Command::new("cargo").arg("metadata")
+        .arg("--no-deps")
+        .output()?;
     if output.status.success() {
-        // None of the unwraps should fail if output of `cargo read-manifest` is correct
         let data = &String::from_utf8(output.stdout).unwrap();
         let json = Json::from_str(data).unwrap();
-        let jtargets = json.find("targets")
+        let mut hitlist: HashSet<&String> = if !workspace_hitlist.is_all() {
+            HashSet::from_iter(workspace_hitlist.get_some().unwrap())
+        } else {
+            HashSet::new() // Unused
+        };
+        let members: Vec<&Json> = json.find("packages")
             .unwrap()
             .as_array()
-            .unwrap();
-        for jtarget in jtargets {
-            targets.push(target_from_json(jtarget));
+            .unwrap()
+            .into_iter()
+            .filter(|member| if workspace_hitlist.is_all() {
+                        true
+                    } else {
+                        let member_name = member.find("name")
+                            .unwrap()
+                            .as_string()
+                            .unwrap();
+                        if hitlist.take(&member_name.to_string()).is_some() {
```

Comment:
You can remove the `if` block here and just return `hitlist.take(&member_name.to_string()).is_some()`